### PR TITLE
kubeadm: fix image pull policy integration

### LINKED ISSUE #103101: [failing test] kubeadm-kinder-latest: task-03-init
#### Which jobs are failing:
https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-latest

#### Which test(s) are failing:
kubeadm-init

#### Since when has it been failing:
06-22

#### Testgrid link:
https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-latest

#### Reason for failure:

```
[preflight] Some fatal errors occurred:
	[ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-apiserver:v1.22.0-alpha.3.377_6dd9deea3daa98: output: time="2021-06-22T17:39:50Z" level=fatal msg="pulling image failed: rpc error: code = NotFound desc = failed to pull and unpack image \"k8s.gcr.io/kube-apiserver:v1.22.0-alpha.3.377_6dd9deea3daa98\": failed to resolve reference \"k8s.gcr.io/kube-apiserver:v1.22.0-alpha.3.377_6dd9deea3daa98\": k8s.gcr.io/kube-apiserver:v1.22.0-alpha.3.377_6dd9deea3daa98: not found"
, error: exit status 1
	[ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-controller-manager:v1.22.0-alpha.3.377_6dd9deea3daa98: output: time="2021-06-22T17:39:50Z" level=fatal msg="pulling image failed: rpc error: code = NotFound desc = failed to pull and unpack image \"k8s.gcr.io/kube-controller-manager:v1.22.0-alpha.3.377_6dd9deea3daa98\": failed to resolve reference \"k8s.gcr.io/kube-controller-manager:v1.22.0-alpha.3.377_6dd9deea3daa98\": k8s.gcr.io/kube-controller-manager:v1.22.0-alpha.3.377_6dd9deea3daa98: not found"
, error: exit status 1
	[ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-scheduler:v1.22.0-alpha.3.377_6dd9deea3daa98: output: time="2021-06-22T17:39:50Z" level=fatal msg="pulling image failed: rpc error: code = NotFound desc = failed to pull and unpack image \"k8s.gcr.io/kube-scheduler:v1.22.0-alpha.3.377_6dd9deea3daa98\": failed to resolve reference \"k8s.gcr.io/kube-scheduler:v1.22.0-alpha.3.377_6dd9deea3daa98\": k8s.gcr.io/kube-scheduler:v1.22.0-alpha.3.377_6dd9deea3daa98: not found"
, error: exit status 1
	[ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-proxy:v1.22.0-alpha.3.377_6dd9deea3daa98: output: time="2021-06-22T17:39:51Z" level=fatal msg="pulling image failed: rpc error: code = NotFound desc = failed to pull and unpack image \"k8s.gcr.io/kube-proxy:v1.22.0-alpha.3.377_6dd9deea3daa98\": failed to resolve reference \"k8s.gcr.io/kube-proxy:v1.22.0-alpha.3.377_6dd9deea3daa98\": k8s.gcr.io/kube-proxy:v1.22.0-alpha.3.377_6dd9deea3daa98: not found"
, error: exit status 1
```

#### Anything else we need to know:

all kubeadm jobs are failing against k/k master.

is this failing after we merged?
https://github.com/kubernetes/kubernetes/pull/102901

### PR DESCRIPTION
**Pull request by neolit123**
#### What this PR does / why we need it:

If the user has not specified a pull policy we must assume a default of
v1.PullIfNotPresent.

Add some extra verbose output to help users monitor what policy is
used and what images are skipped / pulled.

Use "fallthrough" and case handle "v1.PullAlways".

Update unit test.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/103101

#### Special notes for your reviewer:

NONE

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by neolit123** [comment]
/cc @vinayakankugoyal
as this is blocking https://github.com/kubernetes/kubeadm/pull/2511

cc @wangyysde @pacoxu

/triage accepted
/kind regression
/priority critical-urgent

**Review by vinayakankugoyal — APPROVED** [review]

**Comment by wangyysde** [comment]
@neolit123 Thank you for fixing the bug.

**Comment by jackfrancis** [comment]
@neolit123 +1 thank you

Just for my curiosity, I'd like to understand why I experienced this symptom *only* on the 2nd control plane node in a 3 control plane + 2 worker node cluster-api configuration:

```
[2021-06-22 21:05:43] kubeadm version: v1.22.0-alpha.3.388+b1263bc8f5f2ef
[2021-06-22 21:05:43] kubectl version: Client Version: v1.22.0-alpha.3.388+b1263bc8f5f2ef
[2021-06-22 21:05:43] kubelet version: Kubernetes v1.22.0-alpha.3.388+b1263bc8f5f2ef
[2021-06-22 21:05:43] +++ [2021-06-22T21:05:43+00:00] running 'kubeadm join phase preflight --ignore-preflight-errors=DirAvailable--etc-kubernetes-manifests'
[2021-06-22 21:05:43] [preflight] Running pre-flight checks
[2021-06-22 21:05:44] [preflight] Reading configuration from the cluster...
[2021-06-22 21:05:44] [preflight] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'
[2021-06-22 21:05:44] [preflight] Running pre-flight checks before initializing the new control plane instance
[2021-06-22 21:05:44] [preflight] Pulling images required for setting up a Kubernetes cluster
[2021-06-22 21:05:44] [preflight] This might take a minute or two, depending on the speed of your internet connection
[2021-06-22 21:05:44] [preflight] You can also perform this action in beforehand using 'kubeadm config images pull'
[2021-06-22 21:06:04] error execution phase preflight: [preflight] Some fatal errors occurred:
[2021-06-22 21:06:04]  [ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-apiserver:v1.22.0-alpha.3.379_a3f24e84594654: output: time="2021-06-22T21:05:45Z" level=fatal msg="pulling image failed: rpc error: code = NotFound desc = failed to pull and unpack image \"k8s.gcr.io/kube-apiserver:v1.22.0-alpha.3.379_a3f24e84594654\": failed to resolve reference \"k8s.gcr.io/kube-apiserver:v1.22.0-alpha.3.379_a3f24e84594654\": k8s.gcr.io/kube-apiserver:v1.22.0-alpha.3.379_a3f24e84594654: not found"
[2021-06-22 21:06:04] , error: exit status 1
[2021-06-22 21:06:04]  [ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-controller-manager:v1.22.0-alpha.3.379_a3f24e84594654: output: time="2021-06-22T21:05:46Z" level=fatal msg="pulling image failed: rpc error: code = NotFound desc = failed to pull and unpack image \"k8s.gcr.io/kube-controller-manager:v1.22.0-alpha.3.379_a3f24e84594654\": failed to resolve reference \"k8s.gcr.io/kube-controller-manager:v1.22.0-alpha.3.379_a3f24e84594654\": k8s.gcr.io/kube-controller-manager:v1.22.0-alpha.3.379_a3f24e84594654: not found"
[2021-06-22 21:06:04] , error: exit status 1
[2021-06-22 21:06:04]  [ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-scheduler:v1.22.0-alpha.3.379_a3f24e84594654: output: time="2021-06-22T21:05:47Z" level=fatal msg="pulling image failed: rpc error: code = NotFound desc = failed to pull and unpack image \"k8s.gcr.io/kube-scheduler:v1.22.0-alpha.3.379_a3f24e84594654\": failed to resolve reference \"k8s.gcr.io/kube-scheduler:v1.22.0-alpha.3.379_a3f24e84594654\": k8s.gcr.io/kube-scheduler:v1.22.0-alpha.3.379_a3f24e84594654: not found"
[2021-06-22 21:06:04] , error: exit status 1
[2021-06-22 21:06:04]  [ERROR ImagePull]: failed to pull image k8s.gcr.io/kube-proxy:v1.22.0-alpha.3.379_a3f24e84594654: output: time="2021-06-22T21:05:48Z" level=fatal msg="pulling image failed: rpc error: code = NotFound desc = failed to pull and unpack image \"k8s.gcr.io/kube-proxy:v1.22.0-alpha.3.379_a3f24e84594654\": failed to resolve reference \"k8s.gcr.io/kube-proxy:v1.22.0-alpha.3.379_a3f24e84594654\": k8s.gcr.io/kube-proxy:v1.22.0-alpha.3.379_a3f24e84594654: not found"
[2021-06-22 21:06:04] , error: exit status 1
[2021-06-22 21:06:04] [preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`
[2021-06-22 21:06:04] To see the stack trace of this error execute with --v=5 or higher
[2021-06-22 21:06:04] !!! [2021-06-22T21:06:04+00:00] kubeadm reported failed action(s) for 'kubeadm join phase preflight --ignore-preflight-errors=DirAvailable--etc-kubernetes-manifests'
```

The 1st control plane node and both worker nodes came online fine, only the 2nd control plane node failed kubeadm join. (The 3rd control plane node was never built as the cluster-api controllers were blocked on waiting for the 2nd.)

cc @chewong

**Comment by neolit123** [comment]
> Just for my curiosity, I'd like to understand why I experienced this symptom only on the 2nd control plane node in a 3 control plane + 2 worker node cluster-api configuration:

i have no explanation.
our CI caught the problem during init, and double checking the code it should happen for both init and join (control-plane):

https://github.com/kubernetes/kubernetes/blob/0ae157df392617fdac471bdc403cb0ecc8f433ca/cmd/kubeadm/app/cmd/phases/init/preflight.go#L68
https://github.com/kubernetes/kubernetes/blob/0ae157df392617fdac471bdc403cb0ecc8f433ca/cmd/kubeadm/app/cmd/phases/join/preflight.go#L127

**Comment by jackfrancis** [comment]
> i have no explanation.

Thanks for the quick response.

I stumbled against our 3 control plane node tests failing all day yesterday and suspected this change may have been responsible, but was unable to figure it out. I'd love to start running more HA control plane tests on all kubeadm changes to get E2E signal of both kubeadm init and join gestures. I may start doing that as I observe kubeadm changes manually, and then bug you to add these tests as gating requirements. ;)

**Comment by jackfrancis** [comment]
Here's a representative failure for anyone who is curious: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102863/pull-kubernetes-e2e-capz-ha-control-plane/1407436289787039744

**Comment by neolit123** [comment]
> and then bug you to add these tests as gating requirements. ;)

we have e2e test jobs that are considered release-informing, so this is already coordinated with the release team and there is potential for such a regression to block a .0 release if the kubeadm maintainers ask for it.

**Comment by jackfrancis** [comment]
Got it. Should we promote any of these existing E2E jobs to be PR merge-gating?

**Comment by neolit123** [comment]
it's complicated. only kind jobs are currently eligible for PR blocking.
but:
- kind also skips the image pull completely.
- i don't think the kind jobs are multi-control plane.

maybe, what can be done here is improve the unit tests to catch this, by ensuring action X happens for pull-policy Y.
there is already a mechanism to fake pull commands in checks_test.go, but this would need extension.
PRs / ideas welcome.

**Pull request merged**
